File: Playwright1/tests/30 PageFactory.js
import { AutomationSitePage } from './10a TestAutomationSitePage';
import { TodoPage } from './28 FixtureExamplePage';

export class PageFactory{
    constructor(page){
        this.page = page;
    }

    createPage(pageName){
        switch(pageName.toLowerCase()){
            case 'automationsite':
                return new AutomationSitePage(this.page);
            case 'todo':
                return new TodoPage(this.page);
            default:
                throw new Error(`page type ${pageName} is not incorrect or is not supported`);
        }
    }
}

File: Playwright_CS/PlaywrightTests/30 PageFactory.cs
using Microsoft.Playwright;
using System;

public class PageFactory
{
    private readonly IPage _page;

    public PageFactory(IPage _page) => this._page = _page;

    /// <summary>
    /// Automatically creates any page class that has a constructor accepting IPage.
    /// </summary>
    public T Create<T>() where T : class
    {
        // Dynamically calls the constructor: new T(page)
        return (T)Activator.CreateInstance(typeof(T), _page)!;
    }

    // public String method1()
    // {
    //     return;
    // }
}




// using System.Threading.Tasks;
// using Microsoft.Playwright;

// public class PageFactory
// {
//     private readonly IPage _page;

//     public PageFactory(IPage page) => _page = page;

//     // Returns the specific page instance
//     public T Create<T>() where T : class
//     {
//         return typeof(T) switch
//         {
//             Type t when t == typeof(TodoPage) => new TodoPage(_page) as T,
//             Type t when t == typeof(AutomationSitePage) => new AutomationSitePage(_page) as T,
//             _ => throw new ArgumentException($"Page type {typeof(T).Name} is not supported.")
//         };
//     }
// }

